pkg/cache/cache_test.go

```diff
@@ -789,7 +799,97 @@ func CacheTest(createCacheFunc func(config *rest.Config, opts cache.Options) (ca
 					Eventually(out).Should(Receive(Equal(pod)))
 					close(done)
 				})
+				type selectorsTestCase struct {
+					fieldSelectors map[string]string
+					labelSelectors map[string]string
+					expectedPods   []string
+				}
+				DescribeTable("filter informers at list watch with selectors", func(tc selectorsTestCase) {
+					By("creating the cache")
+					builder := cache.BuilderWithOptions(
+						cache.Options{
+							SelectorsByObject: cache.SelectorsByObject{
+								&kcorev1.Pod{}: {
+									Field: fields.Set(tc.fieldSelectors).AsSelector(),
+									Label: labels.Set(tc.labelSelectors).AsSelector(),
+								},
+							},
+						},
+					)
+					informer, err := builder(cfg, cache.Options{})
+					Expect(err).NotTo(HaveOccurred())
+
+					By("running the cache and waiting for it to sync")
+					go func() {
+						defer GinkgoRecover()
+						Expect(informer.Start(informerCacheCtx)).To(Succeed())
+					}()
+					Expect(informer.WaitForCacheSync(informerCacheCtx)).NotTo(BeFalse())
+
+					gvk := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+					sii, err := informer.GetInformerForKind(context.TODO(), gvk)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(sii).NotTo(BeNil())
+					Expect(sii.HasSynced()).To(BeTrue())
+
+					By("adding an event handler listening for object creation which sends the object to a channel")
+					out := make(chan interface{})
+					addFunc := func(obj interface{}) {
+						out <- obj
+					}
+					sii.AddEventHandler(kcache.ResourceEventHandlerFuncs{AddFunc: addFunc})
 
+					By("verifying the filter out object is not received on the channel")
+					obtainedObjs := []interface{}{}
+					for range tc.expectedPods {
+						var obtainedObj interface{}
+						Eventually(out).Should(Receive(&obtainedObj), "should receive something")
+						obtainedObjs = append(obtainedObjs, obtainedObj)
+					}
+					Consistently(out).ShouldNot(Receive(), "should not receive anything else")
```

Comment:
I think, this test is likely to flake / slightly incorrect.
The events might arrive in a different order than the one you expect, which would make the test fail.
Also, if there are events for objects, that actually should be filtered out, `Eventually(out).Should(Receive...` might read it from the queue and `Consistently(out).ShouldNot(Receive`... will still succeed, although the filter did not work correctly.

Reply:
> I think, this test is likely to flake / slightly incorrect.
> The events might arrive in a different order than the one you expect, which would make the test fail.

The ginkgo matcher `ConsistOf(tc.expectedPods)` don't care about the order so we are safe there.

> Also, if there are events for objects, that actually should be filtered out, Eventually(out).Should(Receive... might read it from the queue and Consistently(out).ShouldNot(Receive... will still succeed, although the filter did not work correctly.

` Eventually(out).Should(Receive.` will not be executed since those not expected objs are not going to be part of the loop, they will make `Consistently(out).ShouldNot(Receive` fail.

Reply:
Ah, sorry I missed a detail in the `for` loop on the first look. I though, the test is always waiting to receive one of the `expectedPods`, but it actually isn't, it's basically only using the length of `expectedPods` to determine how many items to receive and checks that there are no more events left in the channel afterwards.
This is all fine, my mistake. You can resolve this.